### ISSUE TITLE
RoundRobin for domain / subdomain

### ISSUE THREAD

**Issue by mariusgrigaitis** [skynetservices/skydns#147]
Lets say that I have configuration like this:

/skydns/local/domain/
/skydns/local/domain/a -. IP1
/skydns/local/domain/b -. IP2
/skydns/local/domain/c -. IP3
/skydns/local/domain/subdom1/
/skydns/local/domain/subdom1/a -. IP4
/skydns/local/domain/subdom1/b -. IP5
/skydns/local/domain/subdom1/c -. IP6

DIG for subdom1.domain.local.skydns returns IP4,5,6
DIG for domain.local.skydns returns IP1,2,3,4,5,6

How could I make 
DIG for subdom1.domain.local.skydns returnn IP4,5,6
DIG for domain.local.skydns returnn IP1,2,3

?

**Comment by miekg** [comment]
You can't.

If you want this you should change `domain` to `east` (could be anything) for the 1,2,3 case and `west` for the 4,5,6 case. If you want to get *everything* you should dig for `*.*.skydns.local`.

Note the libc is picky about domainnames containing funny characters (see #127) for which I will allow the use of `any` instead of `*`.

**Comment by mariusgrigaitis** [comment]
I think it would be nice if skydns supported RR depth. In this case depth=1

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [skydns] RoundRobin for domain ..." ]
>I think it would be nice if skydns supported RR depth. In this case depth=1

  ?

Sorry, you have to be more clear with what you mean with that.

/Miek

--
Miek Gieben

**Comment by mariusgrigaitis** [comment]
I suggest:

-round-robin=-1 (same as current=true)
-round-robin=0 (same as current=false)
-round-robin=1 (round robin one level deep)

**Comment by miekg** [comment]
You want to limit the recursion skydns does? (This has nothing to do with round-robin per se).

I also think there are ways around this "limitation" by placing your records differently in Etcd (i.e. use a different naming scheme).

**Comment by mariusgrigaitis** [comment]
Basically yes. 

Can you suggest a way to place records in etcd to overcome this issue and have subdomain.domain.skydns domain.skydns with different entries? Without doing something like subdomain1.domain.skydns subdomain2.domain.skydns?

**Comment by miekg** [comment]
Let's go back a step. What do you actually want to accomplish?

**Comment by mariusgrigaitis** [comment]
I have a domain for my service like:
messaging.company.services (HTTP backend, round-robin)

And I have domain for internal service infrastructure:
mongodb.messaging.company.services

And I'm unable to assign different ips for mongodb.messaging.company.services (because of recursion in skydns)

**Comment by miekg** [comment]
I think what you need to do is bring the services to the same level, i.e. http.messaging.company.services and mongodb.messaging.company.services. (at least if you want to use SkyDNS). SkyDNS is *not* a fullblown nameserver.

**Comment by thockin** [comment]
I was about to file the same issue.

I'd like to offer sub-domains that are not considered part of their parent.

E.g. I have foo.local which has an A record and I want bar.foo.local to have a different A record *and not* be part of foo.local's A records.

I could do dummy.foo.local and bar.foo.local if only I could add a CNAME or A for foo.local -> dummy.foo.local.

**Comment by miekg** [comment]
> I'd like to offer sub-domains that are not considered part of their parent.
>
> E.g. I have foo.local which has an A record and I want bar.foo.local to have a different A record 
> and not be part of foo.local's A records.

I see, but as I said before this isn't really how SkyDNS is designed.

> I could do dummy.foo.local and bar.foo.local if only I could add a CNAME or A for foo.local -> dummy.foo.local.

... and then skydns should stop recursing when it sees such a CNAME record? This can work, but it is all becoming rather complex.

Rather than using CNAME we could add a `Copy` element to `msg.Service` that copies the records found under whatever `Host` points to. Having it *not* recursing is then a different problem.

Maybe it is nicer to have a Group attribute which could be used to tell SkyDNS the only group queries together from the same group. The `foo.local` would be group `foo` and the levels below that would have group `foo-1`, meaning that foo-1 records will not be included in the reply. The semantics would then be:

* if there is a group on the level of the query, use that. I.e foo.local -> group foo, bar.foo.local -> different group
* if there isn't a group, don't do this and use the normal behavior.

With such a group thing, you don't need `Copy` nor CNAMEs I think.

**Comment by thockin** [comment]
Regarding CNAME, what if SkyDNS just kept CNAME records in a separate tree?  I know you didn't intend it to be a full DNS server, but if you just had a way to say that CNAMEs are special things, off to the side, that point to other names, I could side-step this issue.

Alternately, your Group field sounds workable, though it is getting pretty complex.  As before, happy to try it out if you have a quick hack you want me to patch in...

**Comment by miekg** [comment]
A separate tree can work, but is also complex. And I still believe you'll have the recursion problem, unless you say "if you hit this other tree", we won't recurse...But this is more opaque then a Group attribute.

**Comment by thockin** [comment]
It's your project, and obviously I am coming at it from an "I just want a
DNS server" approach.  I'm happy to try your group approach, if you like.

**Comment by miekg** [comment]
Yeah :-) Something like Group makes sense. @bketelsen: any strong thoughts on this?

**Comment by bketelsen** [comment]
In France for the weekend.  Will try to look

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [skydns] RoundRobin for domain ..." ]
>In France for the weekend.  Will try to look

lame excuse :)

Enjoy!

**Comment by miekg** [comment]
Another "hack" which sadly won't work because client will do the queries is to looks at the recursion-flag in the query (RD bit). If it is `off` only give a one level reply if it is `on` do the full recursion.

Again: a hack that won't work for real world clients.

**Comment by thockin** [comment]
Part of the problem is that we're trying to map DNS onto a struct that
isn't natively DNS.  I appreciate your help - I think this ability makes
SkyDNS more useful overall, but if you told us to buzz off, it would not be
unsurprising. I am glad you didn't, though :)

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [skydns] RoundRobin for domain ..." ]
>Part of the problem is that we're trying to map DNS onto a struct that
>isn't natively DNS.  I appreciate your help - I think this ability makes
>SkyDNS more useful overall, but if you told us to buzz off, it would not be
>unsurprising. I am glad you didn't, though :)

Hehe :)

During the weekend I've thought about this feature and I think it can work and
we still adhere to the SkyDNS philosophy (whatever that maybe).

I'll wait for Brian to chime in.

**Comment by bketelsen** [comment]
So I don't have a strong implementation preference (the cname tree sounds more flexible for the end-user, but harder to wrap my head around the implications of how to implement) I do agree that this feature would be useful and worth supporting.

Semi-related: If etcd is a limiting factor in so many of our tickets as it seems to be lately, why don't we consider a replicated in-memory store using one of the Go raft impls like we had in SkyDNS 1?

**Comment by miekg** [comment]
> Semi-related: If etcd is a limiting factor in so many of our tickets as it seems to be lately, why don't we
> consider a replicated in-memory store using one of the Go raft impls like we had in SkyDNS 1?

I don't think so. Because even in memory you want a tree like structure (which we had in v1). And this problems exists because of "the special meaning given to subdomains", which clashes with the issue raised here.

**Comment by bketelsen** [comment]
i wish we could survey users somehow.  my guess is that nobody is using the "special meaning of subdomains"

**Comment by miekg** [comment]
Well the recursive stuff *is* heavily used. But yeah, usage patterns would be nice.

**Comment by thockin** [comment]
I use SkyDNS to approximate what I would get if I manually wrote DNS
records.  In other words, I start with the DNS records I want to produce
and I try to coax SkyDNS into producing those records.  It has mostly
worked, though it has taken a fair bit of experimenting and sometimes the
tree structure just gets in the way.  I'm not saying I am a typical SkyDNS
user, but you said "survey users" :)

**Comment by miekg** [comment]
Wait... if you don't want SkyDNS to recurse to deeper levels we can just add a flag that makes it stop doing that, and make it behave more like a "real" DNS server backed with Etcd.

**Comment by mariusgrigaitis** [comment]
Maybe you want to use both. I would like to suggest adding one more level into directory structure of SkyDNS like:

/recursive/tld/domain/subdomain
/stantard/tld/domain/subdomain

Or setting value of {"host": x.x.x.x, "recurse": true/false}?

**Comment by thockin** [comment]
> Wait... if you don't want SkyDNS to recurse to deeper levels we can just
add a flag that makes it stop doing that, and make it behave more like a
"real" DNS server backed with Etcd.

I don't think the current etcd structure would allow the expression of
things like we are expressing today.  Recursion is the way we achieve SRV,
multiple A records, etc.  Real DNS has nodes in the tree that are both
"files" and "directories" at the same time.

e.g. an A record for bar.foo.com and a set of A records for foo.com that is
not inclusive bar.foo.com

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [skydns] RoundRobin for domain ..." ]
>Maybe you want to use both. I would like to suggest adding one more level into directory structure of SkyDNS like:
>
>/recursive/tld/domain/subdomain
>/stantard/tld/domain/subdomain

Yes this basically to choice we have to make.

>Or setting value of {"host": x.x.x.x, "recurse": true/false}?

I propose to use a group attribute, so you know what to include. But the basic
idea is the same.

/Miek

--
Miek Gieben

**Comment by bketelsen** [comment]
If we're making choices, I think the group attribute looks better.

On Wed, Apr 15, 2015 at 3:17 AM, Miek Gieben <notifications@github.com>
wrote:

> [ Quoting <notifications@github.com> in "Re: [skydns] RoundRobin for
> domain ..." ]
> >Maybe you want to use both. I would like to suggest adding one more level
> into directory structure of SkyDNS like:
> >
> >/recursive/tld/domain/subdomain
> >/stantard/tld/domain/subdomain
>
> Yes this basically to choice we have to make.
>
> >Or setting value of {"host": x.x.x.x, "recurse": true/false}?
>
> I propose to use a group attribute, so you know what to include. But the
> basic
> idea is the same.
>
> /Miek
>
> --
> Miek Gieben
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/skynetservices/skydns/issues/147#issuecomment-93232761>
> .
>



-- 

----
Brian Ketelsen
bketelsen <at> gmail <dot> com

**Comment by thockin** [comment]
I agree that a group attribute is better than parallel trees.

**Comment by miekg** [comment]
ok. hoping to get this, and multiple domains in, in the next couple of days

**Comment by mariusgrigaitis** [comment]
Any progress on this issue?

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [skydns] Recursion of dns entri..." ]
>Any progress on this issue?

Progress is slow as I don't have much time. But there is a `group` branch, with
some ideas implemented, i.e. a `Group` function in msg/service.go. Once that is
finished and tested, you can simply call Group on the returned services and it
should do what you want it.

/Miek

--
Miek Gieben

**Comment by miekg** [comment]
Please try the `group` branch and provide feedback. It should work, but I'm interested if it's obvious to use and if my choices make sense.

**Comment by mariusgrigaitis** [comment]
Looking at documentation.

I suggest adding example values for a/b that it would be clear that those entries need to have group value

https://github.com/skynetservices/skydns/blob/group/README.md#groups

**Comment by miekg** [comment]
re: @bketelsen  (late reply)
> i wish we could survey users somehow. my guess is that nobody is using the "special meaning of > subdomains"

This is used in for instance Prometheus to target specific sets of hosts to scrape.

**Comment by thockin** [comment]
I still have not had a chance to test the groups branch, I apologize.

**Comment by miekg** [comment]
no problem.

**Comment by mariusgrigaitis** [comment]
I've tested the group branch, looks like it works like expected.

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [skydns] Recursion of dns entri..." ]
>I've tested the group branch, looks like it works like expected.

cool :-) I took the liberty to already merge it to master :-)

I am concerned about one thing which is cname chaining and groups. You might end
up in a situation where you need an IP, but you can't return that specific IP,
because the group's settings don't allow it. Not sure if this is a problem in
practise. I also haven't found a good testcase for it.

/Miek

--
Miek Gieben

**Comment by mariusgrigaitis** [comment]
Wait a little before merging to master yet. I'm still checking some things.

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [skydns] Recursion of dns entri..." ]
>Wait a little before merging to master yet. I'm still checking some things.

I consider this feature still to be "in-flux", so I don't mind changing it.

/Miek

--
Miek Gieben

**Comment by tuannvm** [comment]
There's a problem if *subdom1.domain.local.skydns* have a lot of record (e.g. 3000 records). In this case, when i query *domain.local.skydns*, it tooks very long time to return IP1,2,3. Hope that we can have *limit recursive level* in the next release. Thanks!
